datalad/interface/run.py

```diff
@@ -243,32 +243,78 @@ def __init__(self, patterns, pwd=None, expand=False):
 
         if patterns is None:
             self._maybe_dot = []
-            self._paths = {"patterns": []}
+            self._paths = {"patterns": [], "sub_patterns": {}}
         else:
             patterns, dots = partition(patterns, lambda i: i.strip() == ".")
             self._maybe_dot = ["."] if list(dots) else []
             self._paths = {
                 "patterns": [relpath(p, start=pwd) if isabs(p) else p
-                             for p in patterns]}
+                             for p in patterns],
+                "sub_patterns": {}}
 
     def __bool__(self):
         return bool(self._maybe_dot or self.expand())
 
     __nonzero__ = __bool__  # py2
 
+    def _get_sub_patterns(self, pattern):
+        """Extract sub-patterns from the leading path of `pattern`.
+
+        The right-most path component is successively peeled off until there
+        are no patterns left.
+        """
+        if pattern in self._paths["sub_patterns"]:
+            return self._paths["sub_patterns"][pattern]
+
+        head, tail = op.split(pattern)
+        if not tail:
+            # Pattern ended with a separator. Take the first directory as the
+            # base.
+            head, tail = op.split(head)
+
+        sub_patterns = []
+        seen_magic = glob.has_magic(tail)
+        while head:
+            new_head, tail = op.split(head)
+            if seen_magic and not glob.has_magic(head):
```

Comment:
I quite dislike how "batteries included" often lack basic docstrings:
```
In [2]: glob.has_magic?
Signature: glob.has_magic(s)
Docstring: <no docstring>
File:      /usr/lib/python2.7/glob.py
Type:      function
```

Reply:
magic is very hard to define.